suggesting a pull-request-template. (#3018)
KeYProject/key: .github/pull_request_template.md

> <!-- Thanks for submitting this pull request for KeY -->

Reviewer: I would also add a note:

> By making this pull request, your contribution will be licensed under GPL2-or-later.

Author: Is this necessary? Contributing to a GPL project seems to imply this automatically, I think.

Reviewer: I am not sure. But the a little difference is, that now the author also acknowledges a re-licensing to GPL3.

Reviewer 2: It's in the [GitHub Terms of Service](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license). I am pretty sure that every user on GitHub needed to accept those.

Reviewer 2: If anything, we can add a link to our license. However, a hard-coded (static) reference is not a good idea, in my opinion.

Author: My vote would be to omit this here if possible.

<!-- Since the project has a strict review policy, please make the -->
<!-- reviewer's job easier by providing the necessary information -->
<!-- in the text below. The comments may remain since they will be -->
<!-- invisible when showing the PR. -->

## Related Issue

<!-- Please remove if this PR is not related to an issue. -->
<!-- Please add number if it is in answer to an issue. -->
This pull request answers to issue #.

## Intended Change
<!-- Please give a brief description of what behaviour changes and 
     why it should be changed. -->

## Reason for the pull request

<!-- one item like: 
     bugfix, new feature, refactoring, repaired workflow, ... -->

> ## Touched elements

Reviewer: I see this as redundant information. The same is tracked by Github under "Files changed".

Author: Not quite as mentioned then. But we can drop that.

<!-- A brief overview of the touched code areas. No need to list 
     files (that is available in github). Mention things like:
     TacletApps in key.core, ANTLR for JML, github config, ... -->
     

> ## Information for the reviewer

Reviewer: Unclear what to tell to the reviewers. 

What would make us a review easier?

    
<!-- Additional information to help the reviewer understand the request -->

## Additional contact(s)

<!-- who apart from yourself is involved in this pull request?
     Use @mentions to refer to them here -->
     
<!-- DRAFT MODE: Please note that on the button to submit this pull
     request you can select between submitting a merge-ready request
     or one in draft mode (still evolving). 
     Please use the draft mode unless you think that your proposal

>      should be brought onto master in the current form. -->

Reviewer: Maybe we want additional information, e.g., to create a good changelog.

https://github.com/stevemao/github-issue-templates/blob/master/checklist2/PULL_REQUEST_TEMPLATE.md

Reviewer: Also required may be the consideration: "The documentation is/will be updated".

Author: I do not really believe in Changelogs. Is that not usually a very obvious reason for merge conflicts?

What does "The documentation is/will be updated" mean? I do not understand this point.

Reviewer: > What does "The documentation is/will be updated" mean? I do not understand this point.
The author adds a note, whether this Pull Request results are also reflected in `key-docs`. 

I think changelog are important for a real project (just look at dafny, z3, cbmc, ...).

<!-- Thanks for submitting this pull request for KeY -->
<!-- Since the project has a strict review policy, please make the -->
<!-- reviewer's job easier by providing the necessary information -->
<!-- in the text below. The comments may remain since they will be -->
<!-- invisible when showing the PR. -->

## Related Issue

<!-- Please remove if this PR is not related to an issue. -->
<!-- Please add number if it is in answer to an issue. -->

> This pull request answers to issue #.

Reviewer 3: ```suggestion
This pull request addresses #.
```

## Intended Change

<!-- Please give a brief description of what behaviour changes and 
     why it should be changed. -->

## Type of pull request

<!--- What types of changes does your code introduce? Put an `x` in the box(es) that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactoring (behaviour should not change or only minimally change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] There are changes to the (Java) code
- [ ] There are changes to the taclet rule base
- [ ] There are changes to the deployment/CI infrastructure (gradle, github, ...)
- [ ] Other: 

> ## Ensuring quality

Reviewer: This section is kind of strange. 

Some of these things are hard to check ("runtime performance"). On some other we can discuss for weeks ("javadoc"). And other things like "I wrote a something in key-docs" are not reflected. 

I would such get rid of these points. They do not bring anyone further, or makes the review easier. 

I think this templating becomes to "German" and remembers me on the stupid ISO27001-driven forms back in my apprenticeship. 

Things that make this review easier is something like: "I needed to refactor the method class#method by adding a new parameter. Following classes are automatically updated w/o further intentions."

Author: I took the idea of a `[ ]` list from a reference you suggested earlier, @wadoon.
I am happy with it and without it. Could you edit it?

> - [ ] I made sure that introduced/changed code has been well documented (javadoc).
> - [ ] I added new test case(s) for new functionality.
> - [ ] I have tested the feature as follows: ...
> - [ ] I have checked that runtime performance has not deteriorated  

Reviewer 3: ```suggestion
- [ ] I made sure that introduced/changed code is well documented (javadoc and inline comments).
- [ ] I made sure that new/changed end-user features are well documented (https://github.com/KeYProject/key-docs).
- [ ] I added new test case(s) for new functionality.
- [ ] I have tested the feature as follows: ...
- [ ] I have checked that runtime performance has not deteriorated.
```

## Additional information and contact(s)

<!-- Add further information to help the reviewer understand the request.
     Leave empty if you are sure the reviewer does not need more
     

>      Who apart from yourself is involved in this pull request?
>      Use @mentions to refer to them here -->

Reviewer 3: ```suggestion
     Who apart from yourself is involved in this pull request?
     Use @mentions to refer to them here.
     Use Co-Authored-By in your git commits if applicable. -->
```

     
<!-- DRAFT MODE: Please note that on the button to submit this pull
     request you can select between submitting a merge-ready request
     or one in draft mode (still evolving). 
     Please use the draft mode unless you think that your proposal
     should be brought onto master in the current form. -->

> The contributions within this pull request will be licensed under GPL2-or-later wihtin KeY.

Reviewer 3: ```suggestion
The contributions within this pull request will be licensed under GPLv2 (only) wihtin KeY.
```
See https://github.com/KeYProject/key/discussions/3212#discussioncomment-6538678

> <!-- Thanks for submitting this pull request for KeY -->
> <!-- Since the project has a strict review policy, please make the -->
> <!-- reviewer's job easier by providing the necessary information -->
> <!-- in the text below. The comments may remain since they will be -->
> <!-- invisible when showing the PR. -->

Reviewer 3: ```suggestion
<!--
Thanks for submitting this pull request for KeY.
Since the project has a strict review policy, please make the
reviewer's job easier by providing the necessary information
in the text below. The comments may remain since they will be
invisible when showing the PR.
-->
```
Easier to read (same content). Should ideally be consistent across the template.
